### PR TITLE
clientv3/balance: fixed flaky balancer tests

### PR DESCRIPTION
**Pull request by lavacat**
- added verification step to indirectly verify that all peers are in balancer subconn list

Please read https://github.com/etcd-io/etcd/blob/main/CONTRIBUTING.md#contribution-flow.

### CODE REVIEW

**Comment by lavacat** [comment]
fixes https://github.com/etcd-io/etcd/issues/14158

**Comment by lavacat** [comment]
Looking at the end of the log from failed run https://github.com/etcd-io/etcd/runs/7049120841?check_suite_focus=true

we see `"msg":"picked","picker"` 10 times as expected. But first 4 times "subconn-size":1. That means that same peer will be picked and `switches` won't be incremented


{"level":"info","msg":"state changed","picker":"picker-error","balancer-id":"ckyp592dedhk","connected":true,"subconn":"0xaa01b00","subconn-size":5,"address":"127.0.0.1:38193","old-state":"CONNECTING","new-state":"READY"}
{"level":"info","msg":"updated picker","picker":"picker-roundrobin-balanced","balancer-id":"ckyp592dedhk","policy":"picker-roundrobin-balanced","subconn-ready":["127.0.0.1:38193 (0xaa01b00)"],"subconn-size":1}
{"level":"debug","msg":"picked","picker":"picker-roundrobin-balanced","address":"127.0.0.1:38193","subconn-index":0,"subconn-size":1}
{"level":"debug","msg":"balancer done","picker":"picker-roundrobin-balanced","address":"127.0.0.1:38193","success":true,"bytes-sent":true,"bytes-received":true}
{"level":"debug","msg":"picked","picker":"picker-roundrobin-balanced","address":"127.0.0.1:38193","subconn-index":0,"subconn-size":1}
{"level":"debug","msg":"balancer done","picker":"picker-roundrobin-balanced","address":"127.0.0.1:38193","success":true,"bytes-sent":true,"bytes-received":true}
{"level":"debug","msg":"picked","picker":"picker-roundrobin-balanced","address":"127.0.0.1:38193","subconn-index":0,"subconn-size":1}
{"level":"debug","msg":"balancer done","picker":"picker-roundrobin-balanced","address":"127.0.0.1:38193","success":true,"bytes-sent":true,"bytes-received":true}
{"level":"debug","msg":"picked","picker":"picker-roundrobin-balanced","address":"127.0.0.1:38193","subconn-index":0,"subconn-size":1}
{"level":"info","msg":"state changed","picker":"picker-roundrobin-balanced","balancer-id":"ckyp592dedhk","connected":true,"subconn":"0xaa01ab0","subconn-size":5,"address":"127.0.0.1:36793","old-state":"CONNECTING","new-state":"READY"}
{"level":"info","msg":"updated picker","picker":"picker-roundrobin-balanced","balancer-id":"ckyp592dedhk","policy":"picker-roundrobin-balanced","subconn-ready":["127.0.0.1:36793 (0xaa01ab0)","127.0.0.1:38193 (0xaa01b00)"],"subconn-size":2}
{"level":"info","msg":"state changed","picker":"picker-roundrobin-balanced","balancer-id":"ckyp592dedhk","connected":true,"subconn":"0xaa01ad0","subconn-size":5,"address":"127.0.0.1:37731","old-state":"CONNECTING","new-state":"READY"}
{"level":"info","msg":"updated picker","picker":"picker-roundrobin-balanced","balancer-id":"ckyp592dedhk","policy":"picker-roundrobin-balanced","subconn-ready":["127.0.0.1:36793 (0xaa01ab0)","127.0.0.1:37731 (0xaa01ad0)","127.0.0.1:38193 (0xaa01b00)"],"subconn-size":3}
{"level":"info","msg":"state changed","picker":"picker-roundrobin-balanced","balancer-id":"ckyp592dedhk","connected":true,"subconn":"0xaa01ac0","subconn-size":5,"address":"127.0.0.1:41573","old-state":"CONNECTING","new-state":"READY"}
{"level":"info","msg":"updated picker","picker":"picker-roundrobin-balanced","balancer-id":"ckyp592dedhk","policy":"picker-roundrobin-balanced","subconn-ready":["127.0.0.1:36793 (0xaa01ab0)","127.0.0.1:37731 (0xaa01ad0)","127.0.0.1:38193 (0xaa01b00)","127.0.0.1:41573 (0xaa01ac0)"],"subconn-size":4}
{"level":"info","msg":"state changed","picker":"picker-roundrobin-balanced","balancer-id":"ckyp592dedhk","connected":true,"subconn":"0xaa01b10","subconn-size":5,"address":"127.0.0.1:36837","old-state":"CONNECTING","new-state":"READY"}
{"level":"info","msg":"updated picker","picker":"picker-roundrobin-balanced","balancer-id":"ckyp592dedhk","policy":"picker-roundrobin-balanced","subconn-ready":["127.0.0.1:36793 (0xaa01ab0)","127.0.0.1:36837 (0xaa01b10)","127.0.0.1:37731 (0xaa01ad0)","127.0.0.1:38193 (0xaa01b00)","127.0.0.1:41573 (0xaa01ac0)"],"subconn-size":5}
{"level":"debug","msg":"balancer done","picker":"picker-roundrobin-balanced","address":"127.0.0.1:38193","success":true,"bytes-sent":true,"bytes-received":true}
{"level":"debug","msg":"picked","picker":"picker-roundrobin-balanced","address":"127.0.0.1:36837","subconn-index":0,"subconn-size":5}
{"level":"debug","msg":"balancer done","picker":"picker-roundrobin-balanced","address":"127.0.0.1:36837","success":true,"bytes-sent":true,"bytes-received":true}
{"level":"debug","msg":"picked","picker":"picker-roundrobin-balanced","address":"127.0.0.1:36793","subconn-index":1,"subconn-size":5}
{"level":"debug","msg":"balancer done","picker":"picker-roundrobin-balanced","address":"127.0.0.1:36793","success":true,"bytes-sent":true,"bytes-received":true}
{"level":"debug","msg":"picked","picker":"picker-roundrobin-balanced","address":"127.0.0.1:41573","subconn-index":2,"subconn-size":5}
{"level":"debug","msg":"balancer done","picker":"picker-roundrobin-balanced","address":"127.0.0.1:41573","success":true,"bytes-sent":true,"bytes-received":true}
{"level":"debug","msg":"picked","picker":"picker-roundrobin-balanced","address":"127.0.0.1:37731","subconn-index":3,"subconn-size":5}
{"level":"debug","msg":"balancer done","picker":"picker-roundrobin-balanced","address":"127.0.0.1:37731","success":true,"bytes-sent":true,"bytes-received":true}
{"level":"debug","msg":"picked","picker":"picker-roundrobin-balanced","address":"127.0.0.1:38193","subconn-index":4,"subconn-size":5}
{"level":"debug","msg":"balancer done","picker":"picker-roundrobin-balanced","address":"127.0.0.1:38193","success":true,"bytes-sent":true,"bytes-received":true}
{"level":"debug","msg":"picked","picker":"picker-roundrobin-balanced","address":"127.0.0.1:36837","subconn-index":0,"subconn-size":5}
{"level":"debug","msg":"balancer done","picker":"picker-roundrobin-balanced","address":"127.0.0.1:36837","success":true,"bytes-sent":true,"bytes-received":true}
--- FAIL: TestRoundRobinBalancedResolvableFailoverFromRequestFail (0.00s)

File: clientv3/balancer/balancer_test.go [hunk 1]
```diff
@@ -92,24 +91,32 @@ func TestRoundRobinBalancedResolvableNoFailover(t *testing.T) {
 				return picked, err
 			}
 
-			prev, switches := "", 0
+			var picked string
+			available := make(map[string]struct{})
+			// cycle through all peers to indirectly verify that balancer subconn list is fully loaded
+			// otherwise we can't reliably count switches in the next step
+			for len(available) < tc.serverCount {
```
Comment (lavacat) [inline clientv3/balancer/balancer_test.go:98]: FYI, technically we only need 2 different peers for switch counting logic to work

File: clientv3/balancer/balancer_test.go [hunk 2]
```diff
@@ -277,6 +285,18 @@ func TestRoundRobinBalancedResolvableFailoverFromRequestFail(t *testing.T) {
...
+	available := make(map[string]struct{})
+	// cycle through all peers to indirectly verify that balancer subconn list is fully loaded
+	// otherwise we can't reliably count switches in the next step
+	for len(available) < serverCount {
+		picked, err = reqFunc(context.Background())
+		if err != nil {
+			t.Fatalf("Unexpected failure %v", err)
+		}
+		available[picked] = struct{}{}
+	}
```
Comment (ahrtr) [inline clientv3/balancer/balancer_test.go:298]: I see exact the same code multiple times, can you get them wrapped in a common function?

Comment (lavacat) [reply clientv3/balancer/balancer_test.go:298]: I've followed the same style as original tests, there is a bunch of duplication.

**Review by ahrtr — APPROVED** [review]
LGTM

Thank you @lavacat

**Comment by ahrtr** [comment]
I saw another issue ,
```
 {"level":"debug","msg":"picked","picker":"picker-roundrobin-balanced","address":"127.0.0.1:39987","subconn-index":0,"subconn-size":1}
{"level":"info","msg":"state changed","picker":"picker-roundrobin-balanced","balancer-id":"cld97ag39nv7","connected":false,"subconn":"0xc000316650","subconn-size":5,"address":"127.0.0.1:39987","old-state":"READY","new-state":"CONNECTING"}
{"level":"info","msg":"updated picker","picker":"picker-roundrobin-balanced","balancer-id":"cld97ag39nv7","policy":"picker-roundrobin-balanced","subconn-ready":[],"subconn-size":0}
{"level":"warn","msg":"balancer failed","error":"rpc error: code = Unavailable desc = transport is closing","picker":"picker-roundrobin-balanced","address":"127.0.0.1:39987","success":false,"bytes-sent":true,"bytes-received":false}
--- FAIL: TestRoundRobinBalancedResolvableFailoverFromServerFail (0.01s)
    balancer_test.go:167: Unexpected failure rpc error: code = Unavailable desc = transport is closing
```


https://github.com/etcd-io/etcd/runs/7293195456?check_suite_focus=true

**Comment by ahrtr** [comment]
Just raised a new issue to track this 
https://github.com/etcd-io/etcd/issues/14216

**Pull request merged**